source/concepts.tex

```diff
@@ -0,0 +1,1249 @@
+%!TEX root = std.tex
+\rSec0[concepts.lib]{Concepts library}
+
+\rSec1[concepts.lib.general]{General}
+
+\pnum
+This Clause describes library components that \Cpp{} programs may use to perform
+compile-time validation of template arguments and perform function dispatch
+based on properties of types. The purpose of these concepts is to establish
+a foundation for equational reasoning in programs.
+
+\pnum
+The following subclauses describe core language concepts, comparison concepts,
+object concepts, and callable concepts as summarized in
+\tref{concepts.lib.summary}.
+
+\begin{libsumtab}{Fundamental concepts library summary}{tab:concepts.lib.summary}
+\ref{concepts.lib.corelang} & Core language concepts & \tcode{<concepts>} \\
+\ref{concepts.lib.compare}  & Comparison concepts    &                    \\
+\ref{concepts.lib.object}   & Object concepts        &                    \\
+\ref{concepts.lib.callable} & Callable concepts      &                    \\
+\end{libsumtab}
+
+\rSec2[concepts.lib.general.equality]{Equality Preservation}
+
+\pnum
+An expression is \term{equality preserving} if, given equal inputs, the expression results in
+equal outputs. The inputs to an expression are the set of the expression's operands. The
+output of an expression is the expression's result and all operands modified by the expression.
+
+\pnum
+Not all input values must be valid for a given expression; e.g., for integers \tcode{a}
+and \tcode{b}, the expression \tcode{a / b} is not well-defined when \tcode{b} is \tcode{0}. This
+does not preclude the expression \tcode{a / b} being equality preserving. The
+\term{domain} of an expression is the set of input values for which the
+expression is required to be well-defined.
+
+\pnum
+Expressions required by this specification to be equality preserving are further
+required to be stable: two evaluations of such an expression with the same input
+objects must have equal outputs absent any explicit intervening modification of
+those input objects.
+\begin{note}
+This requirement allows generic code to reason about the current values of objects
+based on knowledge of the prior values as observed via equality preserving expressions.
+It effectively forbids spontaneous changes to an object, changes to an object from
+another thread of execution, changes to an object as side effects of non-modifying
+expressions, and changes to an object as side effects of modifying a distinct object
+if those changes could be observable to a library function via an equality preserving
+expression that is required to be valid for that object.
+\end{note}
+
+\pnum
+Expressions declared in a \grammarterm{requires-expression} in this document are
```

Comment:
We always want explicit convertibility as well when we require implicit convertibility, which means that we'd use `ConvertibleTo` instead. (But perhaps `ConvertibleTo` is missing a requirement that the conversion is equality-preserving?)

Reply:
Hm... ConvertibleTo has a _simple-requirement_, which contains a `static_cast<>` expression, which we're instructed needs to be equality preserving. Ah, but... see below.

Reply:
We use the term "expressions" in library to refer to what are more precisely "expression patterns" that are required to be valid when program-defined types/values are substituted into the expression pattern. For example, [structure.requirements]/4 states: "Requirements are stated in terms of well-defined **expressions** that define valid terms of the types that satisfy the requirements." I'm (ab)using the term similarly here to refer to the expression patterns declared in simple and compound requirements in a *requires-expression*.

I think our intent is (or should be) to include implicit conversion constraints introduced by a *compound-requirement*, although the question is largely moot since we avoid implicit conversion constraints like the plague and (as @timsong-cpp said) typically require deduction constraints a la `{ E } -> ConvertibleTo<T>;` in preference to `{ E } -> T;`. (I've been telling anyone who will listen since 2015 that implicit conversion constraints are largely useless because they result in expression patterns that aren't composable.)

Reply:
I wonder if in the library preamble we can handwave that when the library uses an implicit conversion constrain we also require explicit conversion (with the same semantics) and equality preservation. And then stump for more help from the core language regarding conversion constraints.

Reply:
I'd prefer to be explicit about conversions. I don't think writing `{ E } -> ConvertibleTo<T>;` is a significant burden over `{ E } -> T;` - not enough so to have the same syntax implicitly result in different *checked* constraints, anyway. I'll file an issue to deal with the interaction of equality-preservation and implicit conversion constraints. It doesn't really matter to Ranges, but I suspect someone will eventually want to specify a library component with such a constraint and it needs to have a well-defined meaning rather than us simply telling them to not do that bad thing.

Reply:
Clarification: today we have to write `E; requires ConvertibleTo<decltype((E)), T>;` to express that constraint properly in terms of the exact type and value category of `E`. Once P1084 is adopted, it can be simplified to `{ E } -> ConvertibleTo<T>;`.

Reply:
@CaseyCarter: I don't buy the "expression pattern" theory for the case of new-style concepts. (I agree regarding old-style concepts.) With concepts in the core language, we no longer have to handwave which expressions we mean:  Concepts show actual code that can be compiled. There are no "patterns" going on beyond what the core language prescribes. To wit: the text is referring to the grammar non-terminal _requires-expression_.
My narrow (editorial) issue here is that "expressions declared in..." sounds broken to me; expressions are never "declared". Further (new concern), it's unclear whether "in" is also intended to cover any subexpression of the top-level expressions in the requirements.
I'd appreciate a clear syntax decomposition of requires-expression to define which expressions are in view here (maybe call them "required expressions" or so). Maybe that's not editorial to start with. @zygoloid?

Reply:
Would replacing "Expressions declared in a *requires-expression* in this document..." with "Expressions that appear in a *simple-requirement* or *compound-requirement* of a *requires-expression* in the standard library..." be an improvement? That seems to address much of your concerns.